geopandas/geodataframe.py

```diff
@@ -126,6 +126,8 @@ class GeoDataFrame(GeoPandasBase, DataFrame):
     # TODO: remove "_crs" in 0.12
     _metadata = ["_crs", "_geometry_column_name"]
 
+    _internal_names_set = DataFrame._internal_names_set | {"geometry"}
```

Comment:
I don't know if it matters in practice, but the pandas docs indicate to override both `_internal_names` and `_internal_names_set` (https://pandas.pydata.org/docs/dev/development/extending.html#define-original-properties)

The only place where currently `_internal_names` is used and not `_internal_names_set`, is in unpickling: https://github.com/pandas-dev/pandas/blob/e6024c12f0061602b67f64adcaf8ab5cb77d5418/pandas/core/generic.py#L2157 (and I am not sure we actually want it to be set there)

Reply:
Happy to make that change if it makes sense. Not famililar with the internals but I don't think we need to set `geometry` as part of unpickling - we're not storing unique data on that property, it just retrieves based on `_geometry_column_name`.

Reading over that link again, it almost seems that `geometry` should be added to `_metadata` as it's a "normal" property and not a "temporary property" - but in practice it doesn't seem to work - it doesn't solve the name overlap stuff this PR is intended to address (and it's also causing a whole bunch of other failures).

Reply:
> Reading over that link again, it almost seems that geometry should be added to _metadata as it's a "normal" property and not a "temporary property" - but in practice it doesn't seem to work

I think that doesn't work because `geometry` is an actual property with a setter and getter method defined on the GeoDataFrame. The actual information we need to keep track of is the geometry column name, which is a plain attribute (and that is something that we actually have in `_metadata`)

Let's maybe follow the pandas docs and also add `_internal_names` here, just to be sure?